File: backend/routers/members.py
# Members router for Silent Honor Foundation
import asyncio
from datetime import datetime, timezone
from fastapi import APIRouter, HTTPException, Request, UploadFile, File
from bson import ObjectId

from middleware.auth_middleware import get_current_user
from middleware.logging_middleware import log_audit_event, AUDIT_ACTIONS
from utils.storage import upload_dd214 as storage_upload_dd214
from utils.email import send_admin_notification

router = APIRouter(prefix="/api/member", tags=["Members"])

# Database reference
db = None

def set_db(database):
    global db
    db = database

@router.get("/profile")
async def get_profile(request: Request):
    """Get current member's profile"""
    user = await get_current_user(request)
    return user

@router.put("/profile")
async def update_profile(request: Request):
    """Update member profile"""
    user = await get_current_user(request)
    data = await request.json()

    allowed_fields = ["first_name", "last_name", "phone", "state", "email_preferences"]
    update_data = {k: v for k, v in data.items() if k in allowed_fields}

    if update_data:
        await db.users.update_one(
            {"_id": ObjectId(user["_id"])},
            {"$set": update_data}
        )

    return {"message": "Profile updated successfully"}

@router.get("/counselor")
async def get_assigned_counselor(request: Request):
    """Get member's assigned counselor"""
    user = await get_current_user(request)

    member = await db.users.find_one({"_id": ObjectId(user["_id"])})
    counselor_id = member.get("assigned_counselor_id")

    if not counselor_id:
        return {"id": None, "name": None, "message": "No counselor assigned yet"}

    counselor = await db.users.find_one({"_id": ObjectId(counselor_id)})
    if not counselor:
        return {"id": None, "name": None, "message": "Counselor not found"}

    return {
        "id": str(counselor["_id"]),
        "name": f"{counselor.get('first_name', '')} {counselor.get('last_name', '')}".strip(),
        "email": counselor.get("email"),
        "title": counselor.get("title", "Certified Financial Counselor"),
        "bio": counselor.get("bio", ""),
        "specialties": counselor.get("specialties", []),
        "calendly_url": counselor.get("calendly_url")
    }

@router.get("/courses")
async def get_courses(request: Request):
    """Get member's courses with progress — verified members only"""
    user = await get_current_user(request)

    if not user.get("verified"):
        raise HTTPException(status_code=403, detail="Your account must be verified to access courses.")

    progress = await db.course_progress.find(
        {"user_id": ObjectId(user["_id"])},
        {"course_id": 1, "completed_lessons": 1, "updated_at": 1}
    ).to_list(100)

    # Get courses from database
    db_courses = await db.courses.find({"status": {"$in": ["live", "published", "coming_soon"]}}).sort("created_at", 1).to_list(100)

    courses = []
    for c in db_courses:
        lesson_count = await db.lessons.count_documents({"course_id": str(c["_id"])})
        courses.append({
            "id": str(c["_id"]),
            "title": c.get("title", ""),
            "total_lessons": lesson_count,
            "status": c.get("status", "draft")
        })

    # Merge progress
    progress_map = {str(p.get("course_id")): p for p in progress}
    for course in courses:
        p = progress_map.get(course["id"], {})
        completed = p.get("completed_lessons", [])
        if isinstance(completed, int):
            course["completed_lessons"] = completed
        else:
            course["completed_lessons"] = len(completed)
        course["progress"] = round((course["completed_lessons"] / course["total_lessons"]) * 100) if course["total_lessons"] > 0 else 0

    return courses

@router.post("/courses/{course_id}/progress")
async def update_course_progress(request: Request, course_id: str):
    """Update course progress — verified members only"""
    user = await get_current_user(request)

    if not user.get("verified"):
        raise HTTPException(status_code=403, detail="Your account must be verified to access courses.")

    data = await request.json()

    await db.course_progress.update_one(
        {"user_id": ObjectId(user["_id"]), "course_id": course_id},
        {"$set": {
            "completed_lessons": data.get("completed_lessons", 0),
            "updated_at": datetime.now(timezone.utc)
        }},
        upsert=True
    )

    return {"message": "Progress updated"}

@router.get("/dashboard")
async def get_dashboard_data(request: Request):
    """Get all dashboard data for member"""
    user = await get_current_user(request)
    user_id = ObjectId(user["_id"])

    # Get credit scores
    latest_credit = await db.credit_scores.find_one(
        {"user_id": user_id},
        sort=[("date", -1)]
    )

    # Get disputes count
    disputes_count = await db.disputes.count_documents({"user_id": user_id})
    pending_disputes = await db.disputes.count_documents({"user_id": user_id, "status": {"$in": ["pending", "sent"]}})

    # Get unread messages
    unread_messages = await db.messages.count_documents({"to_user_id": user_id, "read": False})

    # Get course progress
    progress = await db.course_progress.find({"user_id": user_id}).to_list(100)
    courses_in_progress = len([p for p in progress if 0 < p.get("percent_complete", 0) < 100])
    courses_completed = len([p for p in progress if p.get("percent_complete", 0) == 100])

    # Get counselor info
    member = await db.users.find_one({"_id": user_id})
    counselor_assigned = bool(member.get("assigned_counselor_id"))

    return {
        "credit_scores": {
            "equifax": latest_credit.get("equifax") if latest_credit else None,
            "experian": latest_credit.get("experian") if latest_credit else None,
            "transunion": latest_credit.get("transunion") if latest_credit else None,
            "date": latest_credit.get("date").isoformat() if latest_credit and latest_credit.get("date") else None
        },
        "disputes": {
            "total": disputes_count,
            "pending": pending_disputes
        },
        "messages": {
            "unread": unread_messages
        },
        "courses": {
            "in_progress": courses_in_progress,
            "completed": courses_completed
        },
        "counselor_assigned": counselor_assigned,
        "pipeline_stage": member.get("pipeline_stage", "applied"),
        "dd214_status": member.get("dd214_status", "pending")
    }

# DD-214 Upload
@router.post("/dd214")
@router.post("/upload/dd214")
async def upload_dd214(request: Request, file: UploadFile = File(...)):
    """Upload DD-214 document to Supabase or local storage"""
    user = await get_current_user(request)

    # Validate file type
    allowed_types = ["application/pdf", "image/jpeg", "image/jpg", "image/png"]
    if file.content_type not in allowed_types:
        raise HTTPException(status_code=400, detail="Invalid file type. Only PDF, JPG, PNG allowed.")

    # Validate file size (10MB max)
    contents = await file.read()
    if len(contents) > 10 * 1024 * 1024:
        raise HTTPException(status_code=400, detail="File too large. Maximum 10MB.")

    # Upload to storage (Supabase or local fallback)
    result = await storage_upload_dd214(contents, file.filename, user["_id"])

    if not result["success"]:
        raise HTTPException(status_code=500, detail=f"Upload failed: {result.get('error')}")

    filename = result["filename"]
    storage_type = result["storage_type"]

    # Update user record
    await db.users.update_one(
        {"_id": ObjectId(user["_id"])},
        {"$set": {
            "dd214_file": filename,
            "dd214_storage_type": storage_type,
            "dd214_status": "pending_review",
            "dd214_uploaded_at": datetime.now(timezone.utc),
            "pipeline_stage": "dd214_pending"
        }}
    )

    await log_audit_event(
        action=AUDIT_ACTIONS["DD214_UPLOADED"],
        entity_type="user",
        entity_id=user["_id"],
        user_email=user.get("email"),
        details={"storage_type": storage_type},
        ip_address=request.client.host if request.client else None
    )

    # Notify admin of new DD-214 upload
    member_name = f"{user.get('first_name', '')} {user.get('last_name', '')}".strip()
    asyncio.create_task(send_admin_notification(
        "New DD-214 Upload",
        f"{member_name} ({user.get('email')}) has uploaded their DD-214 for verification."
    ))

    return {"message": "File uploaded successfully", "filename": filename, "storage_type": storage_type}

File: backend-rust/crates/api/src/routes/members.rs
//! `/api/member/*` -- profile, counselor, dashboard, courses, course progress.
//!
//! Faithful port of backend/routers/members.py (read + progress endpoints).
//! The DD-214 upload (POST /api/member/dd214) is intentionally deferred to its
//! own PR: it needs S3 + AES-256 encryption and multipart handling.

use std::collections::HashMap;

use axum::extract::{Path, State};
use axum::http::HeaderMap;
use axum::Json;
use bson::oid::ObjectId;
use bson::{doc, Bson, Document};
use serde_json::{json, Value};

use sh_core::{AppError, AppResult};

use super::authenticate;
use crate::state::AppState;

/// Read an integer-ish BSON value (stored as Int32, Int64, or Double) as i64.
fn bson_i64(v: Option<&Bson>) -> Option<i64> {
    match v {
        Some(Bson::Int32(i)) => Some(*i as i64),
        Some(Bson::Int64(i)) => Some(*i),
        Some(Bson::Double(f)) => Some(*f as i64),
        _ => None,
    }
}

// GET /api/member/profile
pub async fn get_profile(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let (uid, user) = authenticate(&state, &headers).await?;
    Ok(Json(user.to_profile(&uid)))
}

// PUT /api/member/profile
pub async fn update_profile(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let (uid, _user) = authenticate(&state, &headers).await?;

    let allowed = [
        "first_name",
        "last_name",
        "phone",
        "state",
        "email_preferences",
    ];
    let mut set = Document::new();
    if let Some(obj) = body.as_object() {
        for key in allowed {
            if let Some(v) = obj.get(key) {
                if let Ok(b) = bson::to_bson(v) {
                    set.insert(key, b);
                }
            }
        }
    }

    if !set.is_empty() {
        let oid = ObjectId::parse_str(&uid).map_err(|_| AppError::Unauthorized)?;
        state
            .db
            .collection::<Document>("users")
            .update_one(doc! { "_id": oid }, doc! { "$set": set })
            .await?;
    }
    Ok(Json(json!({ "message": "Profile updated successfully" })))
}

// GET /api/member/counselor
pub async fn counselor(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let (_uid, user) = authenticate(&state, &headers).await?;

    let counselor_id = match user.assigned_counselor_id.as_deref() {
        Some(c) if !c.is_empty() => c.to_string(),
        _ => {
            return Ok(Json(
                json!({ "id": null, "name": null, "message": "No counselor assigned yet" }),
            ))
        }
    };

    let coid = match ObjectId::parse_str(&counselor_id) {
        Ok(o) => o,
        Err(_) => {
            return Ok(Json(
                json!({ "id": null, "name": null, "message": "Counselor not found" }),
            ))
        }
    };

    let counselor = state
        .db
        .collection::<Document>("users")
        .find_one(doc! { "_id": coid })
        .await?;
    let Some(c) = counselor else {
        return Ok(Json(
            json!({ "id": null, "name": null, "message": "Counselor not found" }),
        ));
    };

    let first = c.get_str("first_name").unwrap_or("");
    let last = c.get_str("last_name").unwrap_or("");
    let specialties: Vec<String> = c
        .get_array("specialties")
        .map(|a| {
            a.iter()
                .filter_map(|b| b.as_str().map(str::to_string))
                .collect()
        })
        .unwrap_or_default();

    Ok(Json(json!({
        "id": coid.to_hex(),
        "name": format!("{first} {last}").trim(),
        "email": c.get_str("email").ok(),
        "title": c.get_str("title").unwrap_or("Certified Financial Counselor"),
        "bio": c.get_str("bio").unwrap_or(""),
        "specialties": specialties,
        "calendly_url": c.get_str("calendly_url").ok(),
    })))
}

// GET /api/member/dashboard
pub async fn dashboard(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let (uid, user) = authenticate(&state, &headers).await?;
    let oid = ObjectId::parse_str(&uid).map_err(|_| AppError::Unauthorized)?;

    let latest_credit = state
        .db
        .collection::<Document>("credit_scores")
        .find_one(doc! { "user_id": oid })
        .sort(doc! { "date": -1 })
        .await?;
    let (equifax, experian, transunion, date) = match &latest_credit {
        Some(d) => (
            bson_i64(d.get("equifax")),
            bson_i64(d.get("experian")),
            bson_i64(d.get("transunion")),
            d.get_datetime("date").ok().map(|dt| dt.timestamp_millis()),
        ),
        None => (None, None, None, None),
    };

    let disputes = state.db.collection::<Document>("disputes");
    let disputes_total = disputes.count_documents(doc! { "user_id": oid }).await?;
    let disputes_pending = disputes
        .count_documents(doc! { "user_id": oid, "status": { "$in": ["pending", "sent"] } })
        .await?;

    let unread = state
        .db
        .collection::<Document>("messages")
        .count_documents(doc! { "to_user_id": oid, "read": false })
        .await?;

    let mut in_progress = 0i64;
    let mut completed = 0i64;
    let mut cursor = state
        .db
        .collection::<Document>("course_progress")
        .find(doc! { "user_id": oid })
        .await?;
    while cursor.advance().await? {
        let d = cursor.deserialize_current()?;
        let pct = bson_i64(d.get("percent_complete")).unwrap_or(0);
        if pct > 0 && pct < 100 {
            in_progress += 1;
        } else if pct >= 100 {
            completed += 1;
        }
    }

    let counselor_assigned = user
        .assigned_counselor_id
        .as_ref()
        .map(|s| !s.is_empty())
        .unwrap_or(false);

    Ok(Json(json!({
        "credit_scores": { "equifax": equifax, "experian": experian, "transunion": transunion, "date": date },
        "disputes": { "total": disputes_total, "pending": disputes_pending },
        "messages": { "unread": unread },
        "courses": { "in_progress": in_progress, "completed": completed },
        "counselor_assigned": counselor_assigned,
        "pipeline_stage": user.pipeline_stage.clone().unwrap_or_else(|| "applied".to_string()),
        "dd214_status": user.dd214_status.clone().unwrap_or_else(|| "pending".to_string()),
    })))
}

// GET /api/member/courses
pub async fn courses(State(state): State<AppState>, headers: HeaderMap) -> AppResult<Json<Value>> {
    let (uid, user) = authenticate(&state, &headers).await?;
    if !user.verified {
        return Err(AppError::Forbidden(
            "Your account must be verified to access courses.".to_string(),
        ));
    }
    let oid = ObjectId::parse_str(&uid).map_err(|_| AppError::Unauthorized)?;

    // Map course_id -> completed lesson count (stored as an int or an array).
    let mut progress: HashMap<String, i64> = HashMap::new();
    let mut pc = state
        .db
        .collection::<Document>("course_progress")
        .find(doc! { "user_id": oid })
        .await?;
    while pc.advance().await? {
        let d = pc.deserialize_current()?;
        let cid = match d.get("course_id") {
            Some(Bson::String(s)) => s.clone(),
            Some(Bson::ObjectId(o)) => o.to_hex(),
            _ => continue,
        };
        let done = match d.get("completed_lessons") {
            Some(Bson::Int32(i)) => *i as i64,
            Some(Bson::Int64(i)) => *i,
            Some(Bson::Array(a)) => a.len() as i64,
            _ => 0,
        };
        progress.insert(cid, done);
    }

    let lessons = state.db.collection::<Document>("lessons");
    let mut out: Vec<Value> = Vec::new();
    let mut cursor = state
        .db
        .collection::<Document>("courses")
        .find(doc! { "status": { "$in": ["live", "published", "coming_soon"] } })
        .sort(doc! { "created_at": 1 })
        .await?;
    while cursor.advance().await? {
        let c = cursor.deserialize_current()?;
        let cid = c
            .get_object_id("_id")
            .map(|o| o.to_hex())
            .unwrap_or_default();
        let total = lessons
            .count_documents(doc! { "course_id": cid.as_str() })
            .await? as i64;
        let done = *progress.get(&cid).unwrap_or(&0);
        let pct = if total > 0 {
            ((done as f64 / total as f64) * 100.0).round() as i64
        } else {
            0
        };
        out.push(json!({
            "id": cid,
            "title": c.get_str("title").unwrap_or(""),
            "total_lessons": total,
            "status": c.get_str("status").unwrap_or("draft"),
            "completed_lessons": done,
            "progress": pct,
        }));
    }

    Ok(Json(json!(out)))
}

// POST /api/member/courses/{course_id}/progress
pub async fn update_progress(
    State(state): State<AppState>,
    Path(course_id): Path<String>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let (uid, user) = authenticate(&state, &headers).await?;
    if !user.verified {
        return Err(AppError::Forbidden(
            "Your account must be verified to access courses.".to_string(),
        ));
    }
    let oid = ObjectId::parse_str(&uid).map_err(|_| AppError::Unauthorized)?;

    let completed = body
        .get("completed_lessons")
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    state
        .db
        .collection::<Document>("course_progress")
        .update_one(
            doc! { "user_id": oid, "course_id": course_id.as_str() },
            doc! { "$set": { "completed_lessons": completed, "updated_at": bson::DateTime::now() } },
        )
        .upsert(true)
        .await?;

    Ok(Json(json!({ "message": "Progress updated" })))
}

// GET /api/member/financial-intake
// Returns the member's saved financial profile (income/expenses/debts/savings/
// goals), or an empty object if they have not filled it in yet.
pub async fn get_financial_intake(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let (uid, _user) = authenticate(&state, &headers).await?;
    let oid = ObjectId::parse_str(&uid).map_err(|_| AppError::Unauthorized)?;

    let found = state
        .db
        .collection::<Document>("financial_intake")
        .find_one(doc! { "member_id": oid })
        .await?;

    match found {
        Some(mut d) => {
            d.remove("_id");
            d.remove("member_id");
            d.remove("updated_at");
            Ok(Json(serde_json::to_value(&d).unwrap_or_else(|_| json!({}))))
        }
        None => Ok(Json(json!({}))),
    }
}

// POST /api/member/financial-intake
// The member submits their own financial profile (from the My Plan intake).
// Stored per member and available to their assigned counselor.
pub async fn save_financial_intake(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> AppResult<Json<Value>> {
    let (uid, _user) = authenticate(&state, &headers).await?;
    let oid = ObjectId::parse_str(&uid).map_err(|_| AppError::Unauthorized)?;

    let mut fields = bson::to_document(&body)
        .map_err(|_| AppError::BadRequest("Invalid financial profile".to_string()))?;
    fields.insert("member_id", oid);
    fields.insert("updated_at", bson::DateTime::now());

    state
        .db
        .collection::<Document>("financial_intake")
        .update_one(doc! { "member_id": oid }, doc! { "$set": fields })
        .upsert(true)
        .await?;

    Ok(Json(json!({ "message": "Financial profile saved" })))
}
